backend/controller/ingress_test.go

```diff
@@ -0,0 +1,86 @@
+package controller
+
+import (
+	"testing"
+
+	"github.com/alecthomas/errors"
+)
+
+func TestMatchURL(t *testing.T) {
+	tests := []struct {
+		pattern  string
+		urlPath  string
+		expected bool
+	}{
+		{"", "", true},
+		{"/", "/", true},
+		{"/users", "/users", true},
+		{"/users/{id}", "/users/123", true},
+		{"/users/{id}/posts", "/users/123/posts", true},
+		{"/users/{id}/posts/{postID}", "/users/123/posts/456", true},
+		{"/users/{id}/posts/{postID}", "/users/123/posts", false},
+		{"/users/{id}/posts/{postID}", "/users/123/posts/456/comments", false},
+		{"/users/{id}/posts/{postID}/comments", "/users/123/posts/456/comments", true},
+		{"/users/{id}/posts/{postID}/comments/{commentID}", "/users/123/posts/456/comments/789", true},
+		{"/users/{id}/posts/{postID}/comments/{commentID}", "/users/123/posts/456/comments", false},
+		{"/users/{id}/posts/{postID}/comments/{commentID}", "/users/123/posts/456/comments/789/replies", false},
+		{"/users/{id}/posts/{postID}/comments/{commentID}/replies", "/users/123/posts/456/comments/789/replies", true},
+		{"/users/{id}/posts/{postID}/comments/{commentID}/replies/{replyID}", "/users/123/posts/456/comments/789/replies/987", true},
+		{"/users/{id}/posts/{postID}/comments/{commentID}/replies/{replyID}", "/users/123/posts/456/comments/789/replies", false},
+		{"/users/{id}/posts/{postID}/comments/{commentID}/replies/{replyID}", "/users/123/posts/456/comments/789/replies/987/extra", false},
+	}
+
+	for _, test := range tests {
+		actual := matchURL(test.pattern, test.urlPath)
+		if actual != test.expected {
```

Comment:
Use `assert.Equal()`

Reply:
There's optional printf-style formatting at the end that you can use to display a nice error eg.

```
assert.Equal(t, test.expected, actual, "pattern = %s, urlPath = %s", test.pattern, test.urlPath)
```